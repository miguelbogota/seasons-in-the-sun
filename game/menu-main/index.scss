@use './menu-main' as menu-main;

@mixin apply {
  @include menu-main.apply;
}
