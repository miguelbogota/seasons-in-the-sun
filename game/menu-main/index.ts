export { MenuMain } from './menu-main.component';
