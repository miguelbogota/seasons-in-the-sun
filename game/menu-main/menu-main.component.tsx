'use client';

import { emitCameraControlEvent } from '@app/game/player/camera-control';
import { Authentication } from '@app/website/authentication';

import { useGameState } from '../state';

/**
 * Component with the main menu of the game.
 */
export function MenuMain() {
  const state = useGameState();

  return (
    !state.isPlaying &&
    !state.isPaused && (
      <div className="menu-main">
        <img src="/images/menu-main-background.webp" alt="Background Image" />

        <div className="menu">
          <h1>Seasons in the Sun</h1>

          <button
            disabled={state.isLoading}
            onClick={() => {
              state.startGame().then(() => {
                setTimeout(() => {
                  emitCameraControlEvent('lock-controls', null);
                }, 200);
              });
            }}
          >
            {state.isLoading ? 'Loading...' : 'Begin'}
          </button>
          <button>Options</button>
        </div>

        <div className="authentication-menu">
          <Authentication />
        </div>
      </div>
    )
  );
}
