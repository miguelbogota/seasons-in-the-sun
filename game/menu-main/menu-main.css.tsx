import { createVar, globalStyle, style, theme } from '@app/theme';

export const root = style({
  width: '100vw',
  height: '100vh',
  position: 'relative',
});

export const image = style({
  selectors: {
    [`${root} > &`]: {
      position: 'absolute',
      top: 0,
      left: 0,
      width: '100vw',
      height: '100vh',
      objectFit: 'cover',
      zIndex: -1,
    },
  },
});

const menuSpace = createVar();

export const menu = style({
  vars: {
    [menuSpace]: '50px',
  },
  selectors: {
    [`${root} > &`]: {
      position: 'absolute',
      left: menuSpace,
      bottom: menuSpace,
    },
  },
});

export const title = style({
  selectors: {
    [`${menu} > &`]: {
      fontSize: '3rem',
      fontWeight: '700',
      marginBottom: '2rem',
      textTransform: 'uppercase',
    },
  },
});

export const button = style({
  selectors: {
    [`${menu} > &`]: {
      display: 'block',
      width: '100%',
      padding: '1rem',
      fontSize: '1.5rem',
      fontWeight: '700',
      textTransform: 'uppercase',
      color: 'white',
      backgroundColor: theme.palette.game('primary'),
      border: 'none',
      borderRadius: '0.5rem',
      marginBottom: '1rem',
      cursor: 'pointer',
      transition: 'background-color 0.2s ease-in-out',
    },
  },
});

globalStyle(`html:has(${root}), body:has(${root})`, {
  // Disables window scrollbars to prevent the page from scrolling while playing.
  // This also prevents the white space when the scroll reaches the end of the page.
  overflow: 'hidden',
});
