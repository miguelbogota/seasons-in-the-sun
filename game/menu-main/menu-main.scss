@mixin apply {
  .menu-main {
    --menu-spacing: 50px;

    width: 100vw;
    height: 100vh;
    position: relative;

    & > img {
      position: absolute;
      top: 0;
      left: 0;
      width: 100vw;
      height: 100vh;
      object-fit: cover;
      z-index: -1;
    }

    & > .menu {
      position: absolute;
      left: var(--menu-spacing);
      bottom: var(--menu-spacing);

      & > h1 {
        font-size: 3rem;
        font-weight: 700;
        margin-bottom: 2rem;
        text-transform: uppercase;
      }

      & > button {
        display: block;
        width: 100%;
        padding: 1rem;
        font-size: 1.5rem;
        font-weight: 700;
        text-transform: uppercase;
        color: white;
        background-color: var(--primary);
        border: none;
        border-radius: 0.5rem;
        margin-bottom: 1rem;
        cursor: pointer;
        transition: background-color 0.2s ease-in-out;
      }
    }

    & > .authentication-menu {
      position: absolute;
      right: var(--menu-spacing);
      top: var(--menu-spacing);
    }

    html:has(&),
    body:has(&) {
      overflow: hidden;
    }
  }
}
