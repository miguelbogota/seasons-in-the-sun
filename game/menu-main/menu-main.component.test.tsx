import { render } from '@testing-library/react';

import { MenuMain } from './menu-main.component';

describe('Game / MenuMain / Component', () => {
  beforeAll(() => {
    global.GUN = () => ({
      // @ts-expect-error Mocking get method.
      get: () => ({}),
      user: () => ({
        // @ts-expect-error Mocking get method.
        recall: () => ({}),
      }),
      on: () => ({}),
    });
  });

  it('should match snapshot', () => {
    const { container } = render(<MenuMain />);
    expect(container).toMatchSnapshot();
  });
});
