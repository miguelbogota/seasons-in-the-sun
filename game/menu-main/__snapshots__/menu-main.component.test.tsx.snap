// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game / MenuMain / Component > should match snapshot 1`] = `
<div>
  <div
    class="menu-main__8gkvil1"
  >
    <img
      alt="Background Image"
      class="menu-main__8gkvil2"
      src="/images/menu-main-background.webp"
    />
    <div
      class="menu-main__8gkvil3"
    >
      <h1
        class="menu-main__8gkvil4"
      >
        Seasons in the Sun
      </h1>
      <button
        class="menu-main__8gkvil5"
      >
        Begin
      </button>
      <button
        class="menu-main__8gkvil5"
      >
        Options
      </button>
    </div>
    <div
      class="menu-main__8gkvil6"
    >
      <div
        class="authentication_root__11zwokp0"
      >
        <pre>
          null
        </pre>
        <button>
          login
        </button>
      </div>
    </div>
  </div>
</div>
`;
